rfcs/20240312-tanop.md

```diff
@@ -0,0 +1,67 @@
+# [RFC] Add TanOp to the StableHLO specification
+
+Status: Review<br/>
+Initial version: 03/12/2024<br/>
+Last updated: 03/15/2024<br/>
+Discussion thread: [GitHub](https://github.com/openxla/stablehlo/pull/2101)
+
+## Motivation
+
+Several features have been added to MHLO in the past year, which frameworks want
+to leverage and members of the community have made requests for them as well.
+This includes `TanOp`.
+
+Frameworks and Compilers both want `tan` op. Jax has [`jnp.tan`](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.tan.html),
+PyTorch has [`torch.tan`](https://pytorch.org/docs/stable/generated/torch.tan.html).
+On Compilers side, XLA has [`mhlo.tan`](https://github.com/tensorflow/mlir-hlo/blob/master/mhlo/IR/hlo_ops.td#L633).
+StableHLO doesn't support `tan` op, but there is an open ticket which requests
+adding this feature in
+[#1358](https://github.com/openxla/stablehlo/issues/1358)
+
+## Proposed Specification
+
+### tan
+
+#### Semantics
+
+Performs element-wise tangent operation on `operand` tensor and
+produces a `result` tensor. Depending on the element type, does the following:
+
+* For floats: `tan` from IEEE-754.
+* For complex numbers: complex tangent.
+* For quantized types:
+  * `dequantize_op_quantize(tan, operand, type(result))`.
+
+#### Inputs
+
+| Label | Name      | Type                                                                    | Constraints |
+|-------|-----------|-------------------------------------------------------------------------|-------------|
+| (I1)  | `operand` | tensor of floating-point or complex type or per-tensor quantized tensor | (C1)        |
+
+#### Outputs
+
+| Name     | Type                                                                    | Constraints |
+|----------|-------------------------------------------------------------------------|-------------|
+| `result` | tensor of floating-point or complex type or per-tensor quantized tensor | (C1)        |
+
+#### Constraints
+
+* (C1) `baseline_type(operand) = baseline_type(result)`.
+
+#### Examples
+
+```mlir
+// %operand: [-1.0, 0.0, 1.0]
```

Comment:
Nit: using the same inputs for sine and cosine example (I remember writing a bullet point on the [checklist](https://github.com/openxla/stablehlo/blob/main/docs/reference.md#testing-guidelines) to use the highest precision possible for examples as default, but can't find it anymore, but let's still use f64 here):
```
// %operand: [
//            [0.0, 1.57079632],       // [0, pi/2]
//            [3.14159265, 4.71238898] // [pi, 3pi/2]
//           ]
%result = "stablehlo.tan"(%operand) : (tensor<2x2xf64>) -> tensor<2x2xf64>
// %result: [
//           [0.0, 0.0],
//           [0.0, 0.0]
//          ]
```